### ISSUE TITLE
Checksums and signatures

### ISSUE THREAD

**Issue by escondida** [soimort/translate-shell#372]
Thanks for translate-shell! It's an excellent program.

The sha1sum provided at https://www.soimort.org/translate-shell/ doesn't match that of the github tarball for 0.9.6.12: e9959df02300279b46e1bf7bc4739b07891d86b4 vs. 43fdeda4d285e928fa8b3bff8b91f2e753241a15. Furthermore, the PGP signature provided on soimort.org doesn't match either! It's concerning, from a security standpoint.

I'd like to request that checksums (ideally at least sha256sums, to prevent collisions--but eh, that's not as big a deal as the PGP) be provided either as a downloadable file with the github release, or as part of the release notes.

More importantly, I'd like to request that a valid PGP signature (from your original soi@mort.ninja key, or from a new key signed by that key) be added as a downloadable file with the github release. Until at least the latter happens, I'm not really comfortable bumping the version of Arch's package.

**Comment by soimort** [comment]
Hi,
Thanks for maintaining the Arch package.

Just to clarify, the sha1sum string `e9959df02300279b46e1bf7bc4739b07891d86b4` is for the standalone file (which can be downloaded from either https://www.soimort.org/translate-shell/trans or http://git.io/trans, or by clicking the "DOWNLOAD" icon just above the sha1sum on the webpage). It's really NOT related to the github tarball at all. Same for PGP there.

The tarballs on github's release page were actually never checksummed or signed before, as many people including me still trust in the company that operates GitHub (and the HTTPS infrastructure it is running on). I may consider doing that for future releases. Should the lack of checksum / signature be an issue for downstream packaging right now, the standalone file (which does have a valid checksum and signature) may be used instead.

**Comment by escondida** [comment]
Hey! Thanks for the response.

Good to know on the sha1string and signature on the soimort page.

Signing github's tarballs is less about trusting github and more about being able to verify the source of the release. For example, it's possible (though I would hope, and choose to assume, unlikely) for your github account to be compromised; it's considerably less likely for the ssh key to your site, your pgp key, *and* your github account to all be compromised.

Summing them is just about basic verification of download integrity. Granted, that's also provided by the signature, but it never hurts to double up.

**Comment by soimort** [comment]
I agree that download *integrity* is important in all cases, especially when the hosting server / network relays are unreliable. But to "verify the source of the release" (which involves verification of *authenticity*) is a more complex goal, and knowing just one GitHub identity is simply not enough. You need to know who (a trustworthy identity, preferably a real person) to verify against.

Not relevant to this project at all, but from my point of view it's worth mentioning that relying on signatures for verification without relating to trustworthy persons / identities can lead to a false sense of security. Let's assume, in the remarkable scenario that my GitHub account is compromised, the attacker can easily forge a fake PGP public key / checksums to replace the genuine ones I post on my personal website. Note that https://www.soimort.org/translate-shell/ is also hosted on GitHub Pages, so anyone who has access to my account can push anything to that page as well. That is to say, in the worst case you just imagined, providing signatures / checksums won't help to guarantee anything anyway; and even worse, you're lured into trusting the malicious ones — without knowing that my account is hijacked at all.

At the end of the day, there's need of a proof from "me" stating that "I am the authentic me"; but this makes no sense assuming my GitHub is already hijacked (and we never exchanged PGP keys so you don't know whom to verify against).

Of course, I'll continue to provide checksums & signatures for integrity checking purposes (that GitHub or your network isn't broken), but let me stress that it's not supposed to be taken as a dead-serious verification of authenticity. (since I basically host my everything public on GitHub :sunglasses:)

**Comment by escondida** [comment]
Naturally, it's not the end-all, be-all; but it's a good indicator that something's gone sideways if things suddenly don't match up and a new key, unsigned by the old one, is being used to sign releases (or releases cease to be signed after having been signed previously, etc.).

I will say that since your page is hosted in the same spot as your git, it might be a good idea to make your key available via public keyservers; that way, there's at least one piece of important information in the chain that has a source elsewhere (even if it's still difficult to verify without actually meeting).

**Comment by soimort** [comment]
> but it's a good indicator that something's gone sideways if things suddenly don't match up and a new key, unsigned by the old one, is being used to sign releases (or releases cease to be signed after having been signed previously, etc.).

Good point.

> there's at least one piece of important information in the chain that has a source elsewhere 

If people want they can find my public key on the MIT PGP server or Keybase.

will leave this issue open until the next release happens. Meanwhile, if that really matters so much, you can choose to use the signed standalone file (instead of the unsigned .gz archives on GitHub) as mentioned before.
